### PR TITLE
test code was hacky with two different GSON instead of having proper …

### PR DESCRIPTION
**Pull request by gsps1**
…serialization logic.

Added serialization logic and required equals and hashcode implementation required for equality check in test case.

### CODE REVIEW

**Comment by gsps1** [comment]
build running https://builds.cask.co/browse/CDAP-RUT1517-1

File: cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/proto/FilterCodec.java [hunk 1]
```diff
@@ -22,14 +22,17 @@
...
 
 import java.lang.reflect.Type;
 import javax.annotation.Nullable;
 
 /**
- * JSON deserializer for {@link Filter}
+ * JSON codec for {@link Filter}
  */
-public class FilterDeserializer implements JsonDeserializer<Filter> {
+public class FilterCodec implements JsonSerializer<Filter>, JsonDeserializer<Filter> {
```
Comment (albertshau) [inline cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/proto/FilterCodec.java:35]: This deserializer is way too complex and error prone (the existing logic, not the stuff you've added).

We should avoid custom serialization/deserialization when possible, especially for proto classes as they are supposed to represent request/response objects. Having this separate codec is brittle and error prone, which is how this bug came up in the first place. Proto objects really shouldn't be interfaces or abstract classes. They shouldn't use generics that aren't actually generic. ValueFilter is only ever with strings, it should not be generic, it should just use string. RangeFilter is only ever int or long, it should not be generic, it should just use longs.

This codec has to understand Filter and all its implementations in order to be correct, which couples everything together. It is much simpler just to have a concrete class that contains nullable whitelist, blacklist, and range fields.

Comment (albertshau) [reply cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/proto/FilterCodec.java:35]: After poking around some more, it looks like only ValueFilter is ever used, and only ever on namespace. Is the use of the other filters yet to be implemented or am I missing something?

Comment (gsps1) [reply cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/proto/FilterCodec.java:35]: the range filter is used for time range fields such as start, end, running time and duration of runs

Comment (gsps1) [reply cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/proto/FilterCodec.java:35]: yes, that makes sense, its true the existing code is unnecessarily complex.

Comment (albertshau) [reply cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/proto/FilterCodec.java:35]: maybe my IDE doesn't have everything indexed then, I only see the filter getter used in ReportGenerationSpark.ReportSparkHandler.getNamespaceFilterIfExists

Comment (gsps1) [reply cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/proto/FilterCodec.java:35]: its used in `co.cask.cdap.report.ReportGenerationHelper.scala` class

**Comment by sreevatsanraman** [comment]
The current PR improves the existing behavior. Opened a techdebt JIRA to fix it properway https://issues.cask.co/browse/CDAP-13840. Merging this.

**Pull request merged**